theano/compile/function.py

```diff
@@ -185,6 +185,22 @@ def opt_log1p(node):
 
 
     """
+    if isinstance(outputs, dict):
+        output_items = outputs.items()
+
+        output_items_sorted = sorted(output_items)
```

Comment:
This still work on non string. Personally, I'm ok to let this, but @lamblin didn't liked it. Can you assert that all key are int or string? Was there a consensus that I missed on how to handle this? Otherwise the PR seem ok except my small comments. thanks